### ISSUE TITLE
Docker: Error response from daemon: Container command not found or does not exist

### ISSUE THREAD

**Issue by Random-Liu** [kubernetes/kubernetes#35914]
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubelet-gce-e2e-ci/10671
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubelet-cri-gce-e2e-ci/701
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubelet-cri-gce-e2e-ci/732

This flake starts happening from last week. Docker reports `Command not found` even when we are sure that the image has this command. For example:
```
Error response from daemon: Container command 'sh' not found or does not exist."
```

For now, I only see this on GCI.

/cc @kubernetes/sig-node

**Comment by yujuhong** [comment]
```
Oct 28 18:21:30 tmp-node-e2e-1b337bad-gci-dev-55-8872-18-0 docker[752]: time="2016-10-28T18:21:30Z" level=error msg="containerd: start container" error="oci runtime error: could not synchronise with container process: stat /var/lib/rkt/pods/exited-garbage/4d59fbf0-8337-476d-aa0b-8f0647e92969/stage1/rootfs/opt/stage2/gci-mounter/rootfs/sys/fs/cgroup/systemd/docker/69f72a5e8d952449f2aa3f72e8764741ba000701bd2c0471ca16e2cfb18df715: no such file or directory" id=69f72a5e8d952449f2aa3f72e8764741ba000701bd2c0471ca16e2cfb18df715
Oct 28 18:21:30 tmp-node-e2e-1b337bad-gci-dev-55-8872-18-0 docker[752]: time="2016-10-28T18:21:30.860813438Z" level=error msg="Handler for POST /v1.23/containers/69f72a5e8d952449f2aa3f72e8764741ba000701bd2c0471ca16e2cfb18df715/start returned error: Container command 'sh' not found or does not exist."
```

Seems to be related to the gci-mounter. /cc @mtaufen @vishh

**Comment by dchen1107** [comment]
I suspect this regression was introduced by recent gci-mounter changes. @vishh and @jingxu97, could one of you guys take a look?

**Comment by yujuhong** [comment]
Another one: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubelet-gce-e2e-ci/10783

**Comment by jingxu97** [comment]
from build log, one error from the failed test is 
Nov  1 15:09:40.405: INFO: At 2016-11-01 15:09:39 +0000 UTC - event for pod-configmaps-33d05617-a045-11e6-90f8-42010a80001c: {kubelet tmp-node-e2e-4c1e55e9-gci-dev-55-8872-18-0} FailedSync: Error syncing pod, skipping: failed to "StartContainer" for "configmap-volume-test" with RunContainerError: "runContainer: Error response from daemon: Container command '/mt' not found or does not exist.

**Comment by yujuhong** [comment]
@jingxu97 I posted a more detailed error message from docker.log in https://github.com/kubernetes/kubernetes/issues/35914#issuecomment-257379974

**Comment by mtaufen** [comment]
@jingxu97 that almost seems like it couldn't find `/mt` in a container based on the `gcr.io/google_containers/mounttest:0.7` image... doubt it's due to a change in that image though, because that image (at 0.7) has been around since May 19. How long has that error been occurring?

**Comment by Random-Liu** [comment]
> doubt it's due to a change in that image though, because that image (at 0.7) has been around since May 19. How long has that error been occurring?

No, this also happened to busybox and other images.

**Comment by vishh** [comment]
I continued the investigation based on https://github.com/kubernetes/kubernetes/issues/35914#issuecomment-257379974.

Docker (v1.11.1) appears to be propagating all cgroupfs mounts on the host to its containers. This is a bug IMHO. @mrunalp do you happen to know why docker isn't ignoring all cgroup mounts excepting the initial mounts?

There is no easy workaround for this since `rkt fly` by default mounts cgroupfs into pods. cc @philips 

@mtaufen I suspect https://github.com/kubernetes/kubernetes/pull/35821/commits/d526d2b59902f07739d088b3f51448ffa1d72adc might alleviate this issue. Unfortunately, that patch cannot go in just by itself. I'd recommend spinning up a local node with this patch and running node e2e until it fails to see if we can repro these failures locally.

**Comment by vishh** [comment]
The docker issue has been fixed upstream via https://github.com/opencontainers/runc/pull/608. the fix is  available in docker v1.12.
As for docker v1.11 & GCI, I'm planning on trying to make rkt not mount cgroups.

**Comment by vishh** [comment]
I tried probing deeper to see where runc is attempting to access the transient cgroups which results in container startup failure, and so far I haven't been able to identify it. 

Meanwhile, by including opencontainers/runc#608 in GCI, we should not hit the issue posted in https://github.com/kubernetes/kubernetes/issues/35914#issuecomment-257379974

**Comment by vishh** [comment]
Given that https://github.com/kubernetes/kubernetes/pull/35821 is merged, the chances of hitting this issue will be much lower. I'm closing this issue since I'm not taking any further action.
If he hit it any further let's re-open and investigate based on new data.

**Comment by yujuhong** [comment]
Just ran into another failure: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubelet-cri-gce-e2e-ci/1042

From docker.log:
```
Nov 16 08:55:19 tmp-node-e2e-ed6c012d-gci-dev-55-8872-18-0 docker[756]: time="2016-11-16T08:55:19Z" level=error msg="containerd: start container" error="oci runtime error: could not synchronise with container process: stat /var/lib/rkt/pods/exited-garbage/a5ffd86f-5733-4bd3-a1d3-14d8d3cd54e8/stage1/rootfs/opt/stage2/gci-mounter/rootfs/sys/fs/cgroup/systemd/docker/869ecbebfb35724fa2dfb836fdbcff193680aa13d4fcd605ffd7eac633bb8d75: no such file or directory" id=869ecbebfb35724fa2dfb836fdbcff193680aa13d4fcd605ffd7eac633bb8d75
Nov 16 08:55:19 tmp-node-e2e-ed6c012d-gci-dev-55-8872-18-0 docker[756]: time="2016-11-16T08:55:19.809875598Z" level=error msg="Handler for POST /v1.23/containers/869ecbebfb35724fa2dfb836fdbcff193680aa13d4fcd605ffd7eac633bb8d75/start returned error: Container command '/mt' not found or does not exist."
```

**Comment by jingxu97** [comment]
I checked the GCI image version for the test, it is gci-dev-55-8872-18-0 which does not has the docker fix @vishh mentioned [opencontainers/runc#608](https://github.com/opencontainers/runc/pull/608).

So far since there is no failure after upgrading the new gci image gci-dev-56-8977-0-0 (https://github.com/kubernetes/kubernetes/pull/36681), I will close this issue for now
